src/ystdlib/io_interface/ReaderInterface.hpp

```diff
@@ -10,140 +13,85 @@
-     */
-    ErrorCode try_read_exact_length(char* buf, size_t num_bytes);
-    /**
-     * Reads a number of bytes
+     * Reads the given number of bytes from the underlying medium into the given buffer.
      * @param buf
-     * @param num_bytes Number of bytes to read
-     * @param eof_possible If EOF should be possible (without reading any bytes)
-     * @return false if EOF is possible and EOF was hit
-     * @return true on success
+     * @param num_bytes Number of bytes to read.
      */
-    bool read_exact_length(char* buf, size_t num_bytes, bool eof_possible);
+    [[nodiscard]] virtual auto read_exact_length(char* buf, size_t num_bytes) -> ErrorCode;
 
     /**
-     * Tries to read a numeric value from a file
-     * @param value The read value
-     * @return Same as FileReader::try_read_exact_length's return values
+     * @param value Returns the read numeric value.
      */
     template <typename ValueType>
-    ErrorCode try_read_numeric_value(ValueType& value);
-    /**
-     * Reads a numeric value
-     * @param value The read value
-     * @param eof_possible If EOF should be possible (without reading any bytes)
-     * @return false if EOF is possible and EOF was hit
-     * @return true on success
-     */
-    template <typename ValueType>
-    bool read_numeric_value(ValueType& value, bool eof_possible);
+    [[nodiscard]] auto read_numeric_value(ValueType& value) -> ErrorCode;
 
     /**
-     * Tries to read a string
      * @param str_length
-     * @param str The string read
-     * @return Same as ReaderInterface::try_read_exact_length
+     * @param str Returns the string read.
      */
-    ErrorCode try_read_string(size_t str_length, std::string& str);
+    [[nodiscard]] virtual auto read_string(size_t str_length, std::string& str) -> ErrorCode;
+
+    // Methods implementing general stream interface
     /**
-     * Reads a string
-     * @param str_length
-     * @param str The string read
-     * @param eof_possible If EOF should be possible (without reading any bytes)
-     * @return false if EOF is possible and EOF was hit
-     * @return true on success
+     * Seeks from the beginning to the given position.
+     * @param pos
      */
-    bool read_string(size_t str_length, std::string& str, bool eof_possible);
+    [[nodiscard]] virtual auto seek_from_begin(size_t pos) -> ErrorCode = 0;
 
     /**
-     * Seeks from the beginning to the given position
-     * @param pos
+     * Seeks from the current position to the next position by the given offset amount.
+     * @param offset
      */
-    void seek_from_begin(size_t pos);
+    // TODO: https://github.com/y-scope/ystdlib-cpp/issues/50
+    // NOLINTNEXTLINE(misc-include-cleaner)
+    [[nodiscard]] virtual auto seek_from_current(off_t offset) -> ErrorCode = 0;
 
     /**
-     * Gets the current position of the read head
-     * @return Position of the read head
+     * @param pos Returns the current position of the stream pointer.
```

Comment:
Similar to above comment.
```suggestion
     * @param pos Returns the current position of the read pointer.
```